cmake/traccc-compiler-options.cmake

```diff
@@ -10,16 +10,24 @@ include_guard( GLOBAL )
 # Include the helper function(s).
 include( traccc-functions )
 
-# Set up the used C++ standard(s).
-set( CMAKE_CXX_STANDARD 17 CACHE STRING "The (host) C++ standard to use" )
+# Set the language standards to use.
+set( CMAKE_CXX_STANDARD 17 CACHE STRING "The (Host) C++ standard to use" )
+set( CMAKE_CUDA_STANDARD 17 CACHE STRING "The (CUDA) C++ standard to use" )
 
 # Turn on a number of warnings for the "known compilers".
 if( ( "${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU" ) OR
     ( "${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang" ) )
 
    # Basic flags for all major build modes.
-   foreach( mode RELEASE RELWITHDEBINFO MINSIZEREL DEBUG )
-      traccc_add_flag( CMAKE_CXX_FLAGS_${mode} "-Wall" )
-      traccc_add_flag( CMAKE_CXX_FLAGS_${mode} "-Wextra" )
-   endforeach()
+   traccc_add_flag( CMAKE_CXX_FLAGS "-Wall" )
+   traccc_add_flag( CMAKE_CXX_FLAGS "-Wextra" )
+
 endif()
+
+# Set the CUDA architecture to build code for.
+set( CMAKE_CUDA_ARCHITECTURES "52" CACHE STRING
+   "CUDA architectures to build device code for" )
```

Comment:
This is only supported in CMake 3.18 as far as I know, perhaps we should ensure that the minimum CMake version is at least 3.18? I know we require 3.9 elsewhere. Also, is it really a smart idea to set defaults for this? I would rather throw an error if this is not set to a sane value by the user.

Reply:
It's a cache variable that we set here. If we set `CXXGPU30_ARCHITECTURES` to some value (assuming that in the c\+\+30 standard there will be a GPU architecture definition necessary), none of the build would care. Setting this variable for CMake <3.18 has absolutely no effect whatsoever. While with CMake 3.18+ if we don't set it, we get a warning from CMake.

I believe setting a default is a good idea. We've been doing that in VecMem since the beginning as well.

https://github.com/acts-project/vecmem/blob/main/cmake/vecmem-compiler-options-cuda.cmake#L17

If the user wants to override this conservative default, they can do it from the command line with `-DCMAKE_CUDA_ARCHITECTURES="52;60;70"`, or something similar. The point is, with a default set, they just don't have to do it all... the... time...

Reply:
I'm not sure I really follow along here. I suppose `CXXGPU30_ARCHITECTURES` is a typo? Anyway, I don't think silently failing on CMake <3.18 is a good idea at all! That seems dangerous to me. Also, The problem with NVIDIA architectures is that they are not necessarily backwards compatible. Even a conservative back-up configuration may still fail to compile on a more recent GPU. I think it would be safest to rely on the default provided by the NVIDIA compiler (which is the default behaviour, I think), because it can make a far more informed decision.

Reply:
:confused: I guess it will help if I give you a bit of a "matrix" of how CMake and CUDA interact with each other these days. Let's use the following example project:

```
[bash][Legolas]:demo-for-stephen > ls
CMakeLists.txt  demo.cu
[bash][Legolas]:demo-for-stephen > more CMakeLists.txt 

cmake_minimum_required( VERSION 3.10 )
project( demo-for-stephen VERSION 0.0.1 LANGUAGES CUDA )

set( CMAKE_CUDA_ARCHITECTURES "52" CACHE STRING "CUDA architecture(s) to use" )

add_executable( demo demo.cu )
[bash][Legolas]:demo-for-stephen > more demo.cu 

__global__
void demo() {

}

int main() {

   demo<<<1, 100>>>();
   return 0;
}
[bash][Legolas]:demo-for-stephen >
```

Now, let's see what happens with a new CMake and CUDA version out of the box.

```
[bash][Legolas]:demo-for-stephen > source ~/software/cuda/11.4.2/x86_64-ubuntu2004/setup.sh 
Configured CUDA from: /home/krasznaa/software/cuda/11.4.2/x86_64-ubuntu2004
[bash][Legolas]:demo-for-stephen > ~/software/cmake/3.21.3/x86_64-ubuntu2004-gcc9-opt/bin/cmake -S . -B build
-- The CUDA compiler identification is NVIDIA 11.4.120
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /home/krasznaa/software/cuda/11.4.2/x86_64-ubuntu2004/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /data/ssd-1tb/projects/traccc/demo-for-stephen/build
[bash][Legolas]:demo-for-stephen >
```

Yepp, everybody is happy. Let's see what happens when `CMAKE_CUDA_ARCHITECTURES` is not set:

```
[bash][Legolas]:demo-for-stephen > ~/software/cmake/3.21.3/x86_64-ubuntu2004-gcc9-opt/bin/cmake -DCMAKE_CUDA_ARCHITECTURES= -S . -B build
-- The CUDA compiler identification is NVIDIA 11.4.120
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /home/krasznaa/software/cuda/11.4.2/x86_64-ubuntu2004/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Configuring done
CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0104 is not set: CMAKE_CUDA_ARCHITECTURES now detected for NVCC,
  empty CUDA_ARCHITECTURES not allowed.  Run "cmake --help-policy CMP0104"
  for policy details.  Use the cmake_policy command to set the policy and
  suppress this warning.

  CUDA_ARCHITECTURES is empty for target "demo".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  Policy CMP0104 is not set: CMAKE_CUDA_ARCHITECTURES now detected for NVCC,
  empty CUDA_ARCHITECTURES not allowed.  Run "cmake --help-policy CMP0104"
  for policy details.  Use the cmake_policy command to set the policy and
  suppress this warning.

  CUDA_ARCHITECTURES is empty for target "demo".
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /data/ssd-1tb/projects/traccc/demo-for-stephen/build
[bash][Legolas]:demo-for-stephen >
```

CMake gets upset that we didn't set it to anything...

Now let's see what happens with an older version of CMake, but still a new CUDA version.

```
[bash][Legolas]:demo-for-stephen > ~/software/cmake/3.14.7/x86_64-ubuntu2004-gcc9-opt/bin/cmake -S . -B build1
-- The CUDA compiler identification is NVIDIA 11.4.120
-- Check for working CUDA compiler: /home/krasznaa/software/cuda/11.4.2/x86_64-ubuntu2004/bin/nvcc
-- Check for working CUDA compiler: /home/krasznaa/software/cuda/11.4.2/x86_64-ubuntu2004/bin/nvcc -- works
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Configuring done
-- Generating done
-- Build files have been written to: /data/ssd-1tb/projects/traccc/demo-for-stephen/build1
[bash][Legolas]:demo-for-stephen > ~/software/cmake/3.14.7/x86_64-ubuntu2004-gcc9-opt/bin/cmake -DCMAKE_CUDA_ARCHITECTURES= -S . -B build2
-- The CUDA compiler identification is NVIDIA 11.4.120
-- Check for working CUDA compiler: /home/krasznaa/software/cuda/11.4.2/x86_64-ubuntu2004/bin/nvcc
-- Check for working CUDA compiler: /home/krasznaa/software/cuda/11.4.2/x86_64-ubuntu2004/bin/nvcc -- works
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Configuring done
-- Generating done
-- Build files have been written to: /data/ssd-1tb/projects/traccc/demo-for-stephen/build2
[bash][Legolas]:demo-for-stephen >
```

That older version of CMake just does not care about this variable **at all**. It is not possible to influence in this easy way which platforms the build would target.

Now let's try your other point about 5.2 being "too new" of a requirement...

```
[bash][Legolas]:demo-for-stephen > ~/software/cmake/3.21.3/x86_64-ubuntu2004-gcc9-opt/bin/cmake -DCMAKE_CUDA_ARCHITECTURES="35" -S . -B build
-- The CUDA compiler identification is NVIDIA 11.4.120
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /home/krasznaa/software/cuda/11.4.2/x86_64-ubuntu2004/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /data/ssd-1tb/projects/traccc/demo-for-stephen/build
[bash][Legolas]:demo-for-stephen > ~/software/cmake/3.21.3/x86_64-ubuntu2004-gcc9-opt/bin/cmake --build build
[ 50%] Building CUDA object CMakeFiles/demo.dir/demo.cu.o
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
[100%] Linking CUDA executable demo
[100%] Built target demo
[bash][Legolas]:demo-for-stephen >
```

5.2 is the **oldest** standard that you can still target with CUDA 11 without getting a warning. It is the most conservative option that we can go with at this very moment. That's why I use it in all the Acts projects right now.

And the reason that I do set a default value is because I hate seeing CMake warnings when I forget to set a value explicitly...

Reply:
Thanks for the demo, but I really feel like we are talking past each other. My point is not that 5.2 is "too new" of a requirement - my point is that NVIDIA architectures are not ordered in the sense that you can run code compiled for an older version of CUDA on a newer device. Case in point: you can design code for CC 5.2 that simply will not run on a CC 5.3 device. Even within the same major version there is _no guarantee of compatibility_ between CUDA versions. I agree that this is confusing, but this is precisely why it makes no sense at all to try to find any kind of default value for this - it's just not how CUDA works. The user should ALWAYS specify the architecture for which they want to build, that is the only way to make sure that it actually works. Frankly, this flows from your own example as well: the compiler will complain at you if you don't set a value. Setting a default at configuration time which does not guarantee that the code will work and _which will obscure this fact from the user by silencing an error_ is, in my opinion, a really bad idea.

Reply:
This is not true...

As I wrote, we've been building VecMem with this same setting since forever. And it's been working just fine on my RTX 3080.